src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs

```diff
@@ -161,8 +186,8 @@ private unsafe void Monitor(AsyncReadState state)
                 overlappedPointer = state.ThreadPoolBinding.AllocateNativeOverlapped(state.PreAllocatedOverlapped);
                 continueExecuting = Interop.Kernel32.ReadDirectoryChangesW(
                     state.DirectoryHandle,
-                    state.Buffer, // the buffer is kept pinned for the duration of the sync and async operation by the PreAllocatedOverlapped
-                    (uint)state.Buffer.Length,
+                    state.Buffer,
+                    (uint)state.BufferByteLength,
                     _includeSubdirectories,
                     (uint)_notifyFilters,
                     null,
```

Comment:
Instead of catching intermittent `ObjectDisposedException`, we can do manual marshalling by calling SafeHandle.DangerousAddRef / DangerousRelease and abort without throwing exceptions when DangerousAddRef fails.

Also, it is not obvious to me where the `ArgumentNullException` caught below can come from. I guess it is left-over from earlier versions of this code.

Reply:
Ah sorry, you are right. I thought that there is a method on SafeHandle with Try semantics, but there is not actually one.